### ISSUE TITLE
control_flow_collections/type_error_test seems to hang when run on the VM

### ISSUE THREAD

**Issue by a-siva** [dart-lang/sdk#36393]
tools/test.py language_2/control_flow_collections/type_error_test

seems to result in 3 tests running forever

```
out/DebugX64/dart --no-background-compilation --enable-experiment=control-flow-collections,constant-update-2018 --ignore-unrecognized-flags --packages=/usr/local/google/home/asiva/workspace/dart-ws2/sdk/.packages /usr/local/google/home/asiva/workspace/dart-ws2/sdk/out/DebugX64/generated_tests/language_2/control_flow_collections/type_error_test_23.dart
out/DebugX64/dart --no-background-compilation --enable-experiment=control-flow-collections,constant-update-2018 --ignore-unrecognized-flags --packages=/usr/local/google/home/asiva/workspace/dart-ws2/sdk/.packages /usr/local/google/home/asiva/workspace/dart-ws2/sdk/out/DebugX64/generated_tests/language_2/control_flow_collections/type_error_test_24.dart
out/DebugX64/dart --no-background-compilation --enable-experiment=control-flow-collections,constant-update-2018 --ignore-unrecognized-flags --packages=/usr/local/google/home/asiva/workspace/dart-ws2/sdk/.packages /usr/local/google/home/asiva/workspace/dart-ws2/sdk/out/DebugX64/generated_tests/language_2/control_flow_collections/type_error_test_25.dart
```

**Comment by a-siva** [comment]
/cc @kmillikin @aartbik 
The test appears to be a very small test and it is unclear why that should result in these processes allocating so much memory and essentially swapping so much that my machine becomes unresponsive.

**Comment by kmillikin** [comment]
The tests have a compile-time error which it looks like we don't catch and that's a bug.  They also have an infinite for loop which the VM will run because we don't signal the error.

There's probably two issues to fix: the missing type error in Fasta and we should consider avoiding the infinite loop in the test.  /cc @munificent

**Comment by kmillikin** [comment]
@stefantsov we're missing the compile-time error on the assignment in the initializer:

```dart
  var _ = <int>[for (int i = "s";;) 1]; //# 23: compile-time error
```
can you take a look when you get a chance?

**Comment by stefantsov** [comment]
The proposed fix for #36395 (https://dart-review.googlesource.com/c/sdk/+/98333) should also resolve this issue.  The compile-time error is now reported, and we're not running the program with an infinite loop in it.
